docs/standard/io/pipelines.md

```diff
@@ -0,0 +1,342 @@
+
+* The logic processes the buffer to skip what's already processed.
+* `PipeReader.AdvanceTo` is called to tell the `PipeReader` how much data has been consumed and examined.
+
+The reader and writer loops end by calling `Complete`. `Complete` lets the underlying Pipe release the memory it allocated.
+
+### Backpressure and flow control
+
+Ideally, reading and parsing work together:
+
+* The writing thread consumes data from the network and puts it in buffers.
+* The parsing thread is responsible for constructing the appropriate data structures.
+
+Typically, parsing takes more time than just copying blocks of data from the network:
+
+* The reading thread gets ahead of the parsing thread.
+* The reading thread has to either slow down or allocate more memory to store the data for the parsing thread.
+
+For optimal performance, there's a balance between frequent pauses and allocating more memory.
+
+To solve the preceding problem, the `Pipe` has two settings to control the flow of data:
+
+* <xref:System.IO.Pipelines.PipeOptions.PauseWriterThreshold>: Determines how much data should be buffered before calls to <xref:System.IO.Pipelines.PipeWriter.FlushAsync%2A> pauses.
+* <xref:System.IO.Pipelines.PipeOptions.ResumeWriterThreshold>: Determines how much data should be buffered before calls to `PipeWriter.FlushAsync` pauses.
+
+The `ResumeWriterThreshold` controls how much the reader has to observe before writing can resume.
+
+![Diagram with ResumeWriterThreshold and PauseWriterThreshold](./media/pipelines/resume-pause.png)
+
+<xref:System.IO.Pipelines.PipeWriter.FlushAsync%2A?displayProperty=nameWithType>:
+
+* Returns an incomplete `ValueTask<FlushResult>` when the amount of data in the `Pipe` crosses `PauseWriterThreshold`.
+* Completes `ValueTask<FlushResult>` when it becomes lower than `ResumeWriterThreshold`.
+
+Two values are used to prevent rapid cycling, which can occur if one value is used.
+
+### Examples
+
+```csharp
+// The Pipe will start returning incomplete tasks from FlushAsync until
+// the reader examines at least 5 bytes.
+var options = new PipeOptions(pauseWriterThreshold: 10, resumeWriterThreshold: 5);
+var pipe = new Pipe(options);
+```
+
+### PipeScheduler
+
+Typically when using `async` and `await`, asynchronous code resumes on either on a <xref:System.Threading.Tasks.TaskScheduler> or on the current  <xref:System.Threading.SynchronizationContext>.
+
+When doing I/O, it's important to have fine-grained control over where the I/O is performed. This control allows taking advantage of CPU caches effectively. Efficient caching is critical for high-performance apps like web servers. <xref:System.IO.Pipelines.PipeScheduler> provides control over where asynchronous callbacks run. By default:
+
+* The current <xref:System.Threading.SynchronizationContext> is used.
+* If there's no `SynchronizationContext`, it uses the thread pool to run callbacks.
+
+[!code-csharp[](~/samples/snippets/csharp/pipelines/Program.cs?name=snippet)]
```

Comment:
Should we say anything about the existence of `PipeScheduler.ThreadPool` and `PipeScheduler.Inline`? And the fact that `PipeScheduler.ThreadPool` is the default?

Reply:
@halter73  can you suggest text? Something like

`[PipeScheduler.ThreadPool](xref:System.IO.Pipelines.PipeScheduler.ThreadPool)`  is the `<xref:System.IO.Pipelines.PipeScheduler>` implementation that queues callbacks to thread pool. `PipeScheduler.ThreadPool` is the default and generally the best choice.  `[PipeScheduler.Inline](xref:System.IO.Pipelines.PipeScheduler.Inline)`  can cause unintended consequences such as deadlocks.

Reply:
That looks great.